truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java

```diff
@@ -242,7 +269,47 @@ public RubyArray mulObject(RubyArray array, int count) {
                 System.arraycopy(store, 0, newStore, storeLength * n, storeLength);
             }
 
-            return new RubyArray(getContext().getCoreLibrary().getArrayClass(), array.getAllocationSite(), newStore, newStoreLength);
+            return new RubyArray(array.getLogicalClass(), array.getAllocationSite(), newStore, newStoreLength);
+        }
+
+        @Specialization(guards = "isRubyString(arguments[1])")
+        public Object mulObject(VirtualFrame frame, RubyArray array, RubyString string) {
+            notDesignedForCompilation();
+            return ruby(frame, "join(sep)", "sep", string);
+        }
+
+        @Specialization(guards = {"!isRubyString(arguments[1])"})
+        public Object mulObjectCount(VirtualFrame frame, RubyArray array, Object object ) {
+            notDesignedForCompilation();
+            if (respondToToStrNode == null) {
+                CompilerDirectives.transferToInterpreter();
+                respondToToStrNode = insert(KernelNodesFactory.RespondToNodeFactory.create(getContext(), getSourceSection(), new RubyNode[] { null, null, null }));
+            }
+            if(respondToToStrNode.doesRespondTo(frame, object, getContext().makeString("to_str"), false)){
```

Comment:
Given this is `notDesignedForCompilation`, you could even write the condition here using `ruby` - `if ((boolean) ruby(frame, "object.respond_to 'to_str'", "object", object)) {`. The avoids the need for the node. No need to change it - just thought I'd point it out.